File: data/repository/src/main/java/com/chronos/mobile/data/repository/AppStateAssembler.kt
package com.chronos.mobile.data.repository

import com.chronos.mobile.core.model.AppState
import com.chronos.mobile.core.model.Course
import com.chronos.mobile.core.model.Timetable
import com.chronos.mobile.core.model.TimetableDetails
import com.chronos.mobile.core.model.TimetableSummary
import com.chronos.mobile.data.preferences.UserPreferenceState
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class AppStateAssembler @Inject constructor() {
    fun resolveCurrentTimetableId(
        timetables: List<TimetableSummary>,
        preferences: UserPreferenceState,
    ): String? = preferences.currentTimetableId
        ?.takeIf { id -> timetables.any { it.id == id } }
        ?: timetables.firstOrNull()?.id

    fun assemble(
        timetables: List<TimetableSummary>,
        preferences: UserPreferenceState,
        currentTimetableId: String?,
        currentTimetable: Timetable?,
    ): AppState = AppState(
        timetables = timetables.toList(),
        currentTimetableId = currentTimetableId,
        wallpaperUri = preferences.wallpaperUri,
        currentTimetable = currentTimetable?.copyForStateBoundary(),
        themeMode = preferences.themeMode,
        useDynamicColor = preferences.useDynamicColor,
    )
}

private fun Timetable.copyForStateBoundary(): Timetable = copy(
    courses = courses.map(Course::copyForStateBoundary).toList(),
    details = details.copyForStateBoundary(),
)

private fun Course.copyForStateBoundary(): Course = copy(
    weeks = weeks.toList(),
)

private fun TimetableDetails.copyForStateBoundary(): TimetableDetails = copy(
    periodTimes = periodTimes.toList(),
)

File: src/lib/storage/app-state-assembler.ts
import { type AppState, type TimetableSummary } from '$lib/models/app-state';
import type { Timetable } from '$lib/models/timetable';
import { copyForStateBoundary } from './mappers';
import type { UserPreferenceState } from './settings-repo';

export function resolveCurrentTimetableId(
	timetables: TimetableSummary[],
	preferences: UserPreferenceState
): string | null {
	const preferred = preferences.currentTimetableId;
	if (preferred && timetables.some((timetable) => timetable.id === preferred)) {
		return preferred;
	}
	return timetables[0]?.id ?? null;
}

export function assembleAppState(
	timetables: TimetableSummary[],
	preferences: UserPreferenceState,
	currentTimetableId: string | null,
	currentTimetable: Timetable | null,
	wallpaperUri: string | null
): AppState {
	return {
		timetables: [...timetables],
		currentTimetableId,
		wallpaperUri,
		currentTimetable: currentTimetable ? copyForStateBoundary(currentTimetable) : null,
		themeMode: preferences.themeMode
	};
}
